File: dart_sdk/lib/utils/amount_formatter.dart
import 'package:apolopay_sdk/apolopay_sdk.dart';
import 'package:intl/intl.dart';

String amountFormatter(
  dynamic amount, {
  String? symbol,
  int decimals = 8,
  I18nLocale? locale,
}) {
  final language = {
    I18nLocale.es: 'es_ES',
    I18nLocale.en: 'en_US',
  }[locale ?? I18nLocale.es];

  if (amount == null || amount == "") {
    amount = 0;
  }

  double amountNumber = double.tryParse(amount.toString()) ?? 0.0;

  // Truncar decimales sin redondear
  final sign = amountNumber < 0 ? "-" : "";
  final absAmount = amountNumber.abs();

  // Usar toString() para obtener la representación literal preferida de Dart
  // y manejar notación científica si es necesario.
  String absAmountStr = absAmount.toString();
  if (absAmountStr.contains('e')) {
    // Si es notación científica, usamos toStringAsFixed pero con un límite alto
    absAmountStr =
        absAmount.toStringAsFixed(20).replaceAll(RegExp(r'\.?0+$'), "");
  }

  final parts = absAmountStr.split('.');
  String truncatedStr = parts[0];

  if (parts.length > 1 && decimals > 0) {
    String decimalPart = parts[1];
    if (decimalPart.length > decimals) {
      decimalPart = decimalPart.substring(0, decimals);
    }
    truncatedStr = "$truncatedStr.$decimalPart";
  }

  final finalAmount = double.parse("$sign$truncatedStr");

  final formatter = NumberFormat.decimalPattern(language);
  formatter.maximumFractionDigits = decimals;
  formatter.minimumFractionDigits = 0;

  final formattedNumber = formatter.format(finalAmount);

  return symbol != null ? "$formattedNumber $symbol" : formattedNumber;
}

File: javascript-sdk/packages/ui/src/utils/amount-formatter.ts
import type { Locale } from "@apolopay-sdk/core";

export function amountFormatter(
  amount?: number | string | null,
  options: {
    decimals?: number,
    symbol?: string,
    lang?: Locale
  } = {}
): string {
  const { decimals = 8, symbol, lang = 'es' } = options;
  const locale = {
    'es': 'es-ES',
    'en': 'en-US',
    'pt': 'pt-BR'
  }[lang];

  if (amount === undefined || amount === null || amount === "") {
    amount = 0;
  }

  const amountNumber = Number(amount);
  if (isNaN(amountNumber)) return "0";

  const sign = amountNumber < 0 ? "-" : "";
  
  const absAmountStr = Math.abs(amountNumber).toLocaleString('en-US', {
    useGrouping: false,
    maximumFractionDigits: 20,
  });
  
  let truncatedStr = absAmountStr;

  const parts = absAmountStr.split('.');
  if (parts.length > 1) {
    const integerPart = parts[0];
    const decimalPart = parts[1];
    if (decimals === 0) {
      truncatedStr = integerPart;
    } else {
      truncatedStr = `${integerPart}.${decimalPart.slice(0, decimals)}`;
    }
  }

  const finalAmount = Number(`${sign}${truncatedStr}`);

  const formatOptions: Intl.NumberFormatOptions = {
    minimumFractionDigits: 0,
    maximumFractionDigits: decimals,
  };

  try {
    return finalAmount.toLocaleString(locale, {
      ...formatOptions,
      style: symbol ? 'currency' : 'decimal',
      currency: symbol,
    });
  } catch (error) {
    const formattedNumber = finalAmount.toLocaleString(locale, {
      ...formatOptions,
      style: 'decimal',
    });
    return symbol ? `${formattedNumber} ${symbol}` : formattedNumber;
  }
}